src/test/kotlin/org/objectionary/deog/unit/graph/builder/BuilderTest.kt

```diff
@@ -24,45 +24,74 @@
+import org.objectionary.deog.graph.repr.DGraphNode
+import org.objectionary.deog.sources.SrsTransformed
+import org.objectionary.deog.sources.XslTransformer
+import org.objectionary.deog.unit.graph.TestBase
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.CsvSource
+import java.io.ByteArrayOutputStream
+import java.nio.file.Path
 
-class BuilderTest : BuilderBase() {
-    @Test
-    fun `test basic dir`() = doTest()
-
-    @Test
-    fun `test multiple aliases`() = doTest()
-
-    @Test
-    fun `test basic tree`() = doTest()
-
-    @Test
-    fun `test tree`() = doTest()
-
-    @Test
-    fun `test multiple trees`() = doTest()
-
-    @Test
-    fun `test basic cycle`() = doTest()
-
-    @Test
-    fun `test triple cycle`() = doTest()
-
-    @Test
-    fun `test multiple cycles`() = doTest()
-
-    @Test
-    fun `test closed cycle`() = doTest()
+/**
+ * Base class for graph builder testing
+ */
+open class BuilderTest : TestBase {
+    private val postfix = "tmp"
 
-    @Test
-    fun `test multiple closed cycles`() = doTest()
+    @ParameterizedTest
+    @CsvSource(
+        value = [
+            "basic_cycle",
+            "basic_dir",
+            "basic_tree",
+            "closed_cycle",
+            "inner",
+            "inner_concrete",
+            "inner_ordered",
+            "multiple_aliases",
+            "multiple_closed_cycles",
+            "multiple_cycles",
+            "multiple_trees",
+            "tree",
+            "triple_cycle"
+        ], ignoreLeadingAndTrailingWhitespace = true
+    )
+    fun `graph builder test`(testName: String) {
+        val sources = SrsTransformed(constructInPath(testName), XslTransformer(), postfix)
+        val graph = GraphBuilder(sources.walk()).createGraph()
+        val actual = stringOutput(graph.heads)
+        val expected = constructOutPath(testName).toFile().bufferedReader().use { it.readText() }
+        checkOutput(expected, actual, "In test: ${constructInPath(testName)}")
+        deleteTempDir(sources.resPath)
+    }
 
-    @Test
-    fun `test inner`() = doTest()
+    override fun constructOutPath(directoryName: String): Path =
+        Path.of("src${sep}test${sep}resources${sep}unit${sep}out${sep}builder$sep$directoryName.txt")
 
-    @Test
-    fun `test inner concrete`() = doTest()
+    private fun stringOutput(
+        heads: MutableSet<DGraphNode>
+    ): String {
+        val out = ByteArrayOutputStream()
+        heads.sortedBy { it.name }.forEach { printOutRecursive(it, out, mutableSetOf()) }
+        return String(out.toByteArray())
+    }
 
-    @Test
-    fun `test inner ordered`() = doTest()
+    private fun printOutRecursive(
+        node: DGraphNode,
+        out: ByteArrayOutputStream,
+        nodes: MutableSet<DGraphNode>
+    ) {
+        out.write("NODE: name=\"${node.name}\"\n".toByteArray())
+        if (!nodes.contains(node)) {
+            nodes.add(node)
+        } else {
+            return
```

Comment:
@c71n93 Is it a good practice to return nothing?

Reply:
@Graur This function was not written by me. As I understood, this function is recursively prints graph in `ByteArrayOutputStream` and returns nothing. I don't really know, is it bad practice to return nothing?